src/main/scala/io/iohk/ethereum/consensus/validators/BlockWithCheckpointHeaderValidator.scala

```diff
@@ -0,0 +1,168 @@
+    blockHeader: BlockHeader,
+    parentHeader: BlockHeader
+  ): Either[BlockHeaderError, BlockHeaderValid] = {
+    blockHeader.checkpoint
+      .map { checkpoint =>
+        lazy val signaturesWithRecoveredKeys = checkpoint.signatures.map(s => s -> s.publicKey(parentHeader.hash))
+
+        // if at least 2 different signatures came from the same signer it will be in this set (also takes care
+        // of duplicate signatures)
+        lazy val repeatedSigners = signaturesWithRecoveredKeys
+          .groupBy(_._2)
+          .filter(_._2.size > 1)
+          .keySet
+          .flatten
+
+        lazy val (validSignatures, invalidSignatures) = signaturesWithRecoveredKeys.partition {
+          //signatures are valid if the signers are known AND distinct
+          case (sig, Some(pk)) => blockchainConfig.checkpointPubKeys.contains(pk) && !repeatedSigners.contains(pk)
+          case _ => false
+        }
+
+        // we fail fast if there are too many signatures (DoS protection)
+        if (checkpoint.signatures.size > blockchainConfig.checkpointPubKeys.size)
+          Left(HeaderWrongNumberOfCheckpointSignatures(checkpoint.signatures.size))
+        else if (invalidSignatures.nonEmpty) {
+          val sigsWithKeys = invalidSignatures.map {
+            case (sig, maybePk) => (sig, maybePk.map(ByteStringUtils.hash2string))
+          }
+          Left(HeaderInvalidCheckpointSignatures(sigsWithKeys))
+        } else if (validSignatures.size < blockchainConfig.minRequireSignatures)
+          Left(HeaderWrongNumberOfCheckpointSignatures(validSignatures.size))
+        else
+          Right(BlockHeaderValid)
+      }
+      .getOrElse(Left(HeaderUnexpectedError("Attempted to validate a checkpoint on a block without a checkpoint")))
+  }
+
+  /**
+    * Validates emptiness of:
+    * - beneficiary
+    * - extraData
+    * - treasuryOptOut
+    * - ommersHash
+    * - transactionsRoot
+    * - receiptsRoot
+    * - logsBloom
+    * - nonce
+    * - mixHash
+    *
+    * @param blockHeader BlockHeader to validate.
+    * @return BlockHeader if valid, an [[io.iohk.ethereum.consensus.validators.BlockHeaderError.HeaderFieldNotEmptyError]] otherwise
+    */
+  private def validateEmptyFields(blockHeader: BlockHeader): Either[BlockHeaderError, BlockHeaderValid] = {
+    if (blockHeader.beneficiary != BlockHeader.EmptyBeneficiary)
+      notEmptyFieldError("beneficiary")
+    else if (blockHeader.ommersHash != BlockHeader.EmptyOmmers)
+      notEmptyFieldError("ommersHash")
+    else if (blockHeader.transactionsRoot != BlockHeader.EmptyMpt)
+      notEmptyFieldError("transactionsRoot")
+    else if (blockHeader.receiptsRoot != BlockHeader.EmptyMpt)
+      notEmptyFieldError("receiptsRoot")
+    else if (blockHeader.logsBloom != BloomFilter.EmptyBloomFilter)
+      notEmptyFieldError("logsBloom")
+    else if (blockHeader.extraData.nonEmpty)
+      notEmptyFieldError("extraData")
+    else if (blockHeader.treasuryOptOut.isDefined)
```

Comment:
Note that this might change after #709 (although probably this will get merged before that)